packages/cli/src/commands/generate/auth/templates/firebase.auth.js.template

```diff
@@ -22,7 +22,7 @@ const config = {
 
 const adminApp = admin.initializeApp(config)
 
-export const getCurrentUser = async (token) => {
+export const getCurrentUser = async (decoded, { token, authType }) => {
```

Comment:
I can see the use of `decoded` vs `token` to be confusing here given that they are equal in the case of Firebase.

Perhaps the 2nd param should be null here?

Reply:
Ah, so I think for things that use the "custom" auth type we should make the "decoded" value undefined, the user would to skip over the first value, and destructure the raw values.

This would be a breaking change, but would make things consistent in the long run.

Reply:
That makes more sense -- then one isn't switching terms ... token is token.

Reply:
@peterp just so I follow, the "custom" vase you mention isn't "firebase" since here it still uses the `token` the decoded and its value just happen to be the same. 

But for the custom 3rd party case, currently:

```
    // These tokens require a 3rd party library for decoding that we don't want to
    // bundle with each installation. We'll cover it in the documentation.
    default: {
      decoded = {
        type,
        token,
      }
      break
    }
  }

  if (decoded === null) {
    throw new AuthenticationError(
      'The authentication token could not be decoded.'
    )
  }
```

do something like:

```
    // These tokens require a 3rd party library for decoding that we don't want to
    // bundle with each installation. We'll cover it in the documentation.
    default: {
      decoded = null
      break
    }
  }

  if (type != 'custom' && decoded === null) {
    throw new AuthenticationError(
      'The authentication token could not be decoded.'
    )
  }
```

Or perhaps check if `type` is not `custom` but in `SupportedAuthTypes` ?

Reply:
Yup, 100%... 

And, I think perhaps that us throwing an error here is is the wrong place to make this happen. The library that is doing the decoding should throw.

Reply:
Just so I make sure I complete the desired changes:

- [ ] In the default case (aka unsupported) set the `decoded = null`
- [] Since 'custom' is in fact a SupportedAuthTypes (see: typesToClients in authClients), add a case for the the 'custom' case and also  ... hmmm ...

Scratch that thought.

Perhaps instead of each case statement calling some specific code to decode the token, for example the current implementation of:

`decoded = jwt.decode(token)`

or

`decoded = await verifyAuth0Token(token)`

or

`decoded = token`

there should be some method on the AuthClient that can `decodeToken`.

> The library that is doing the decoding should throw.

That library should be the AuthClient, no?

So, add to the interface of an AuthClient not only a getToken()  -- which exists -- but also a decodeToken()? 

And make the AuthClient responsible for its own decoding -- and raise if it fails to decode (or fails to verify)?

And then well ... there's no case/switch statement anymore?

If the type is supported (even 'custom'), get its authClient and decode it and send back the `decoded, {authType, token}`

If for some reason the authType is not supported, then return `decoded = null`?

Reply:
> That library should be the AuthClient, no?
Yup, often we don't install those auth clients because that would bloat our bundles - so right now we only support jwt's either via Netlify or via Auth0.

> And then well ... there's no case/switch statement anymore?
Totallly!

> If for some reason the authType is not supported, then return decoded = null?
Yup, yup, yup.

Reply:
Perfect. It'll add several more changes, but will be a much more extensible solution. 

I imagine the AuthClient interface will see many more updates ... since I could see it being responsible for checking roles and permissions. For Auth0 decode/get the userMetadata and the see if the user `hasRole(role)` or `can(permission)`.

An app's auth could then implement its own `isAdmin()` as being `.hasRole('admin')` or something  ... and `canPublish()` as `can('publish.post'`).

But that's for another day.

I'll work on updating the AuthClient interface and moving the token decodes into each supported client and then refactor.

Reply:
Your imagination is directly aligned with ours! :)

Reply:
@peterp So, the idea that the `AuthClient` can be responsible for decode the token is a good one, but I ran into the issue that often the clients reference `window` and as such cannot be used on the `api` side:

```
api | ReferenceError: window is not defined
```

I made changes so that in authHeaders:

```
// authHeaders

import type { SupportedAuthClients, SupportedAuthTypes } from '@redwoodjs/auth'
import { createAuthClient } from '@redwoodjs/auth'
```

I could then get the client based on type (in its `decodeAuthToken`) ... but that's when the `window` issue happens because of importing the AuthClients.

Ideas how to 1) get around or 2) not implement like this?

Reply:
Actually, that's why when fetching the `userProfile` from Auth0, I used:

`import { AuthenticationClient } from 'auth0'`

not

`import type { Auth0Client as Auth0 } from '@auth0/auth0-spa-js'`

Auth here is the client-side one ... and that won't work on the api side.

Reply:
OK having chatted with @peterp what we really want to do is to cleanup the `switch` and have a nicer shape/pattern to add decoding rules per client.

Something perhaps along these line:

```
const decodeToken = async ({type, token}) => {
  const decoders = {
    'auth0': function (token) {
      return await verifyAuth0Token(token);
    },
    'netlify': function (token) {
      // the prod vs dev rules here
      return decoded
    },
    'goTrue': function (token) {
      // the prod vs dev rules here
      return decoded
    }, 
    'firebase': function (token) {
      return token;
    },    
    'magiclink': function (token) {
      return token;
    },
    'custom': function (token) {
      return {type, token}
    }        
  };
  return decoders[type](token)
}
```

then can call from `decodeAuthToken` in `AuthHeaders`...

```
const decoded = decodeToken({type, token})
```

but maybe better names since `decodeAuthToken` vs `decodeToken` is way too similar and can be confusing.

is this the right direction @peterp ?

can keep that in AuthHeaders or ... move into somewhere in `@redwoodjs/auth`?

Reply:
Actually, I have an idea for a better implementation that follows the AuthClient pattern more closely.

Reply:
@peterp I've refactored to have `AuthDecoders` defined similar to the way the `AuthClients` are.

Each AuthDecoder implements an interface that can `decodeToken()` and provides what it needs (typically the event, context).

Important: I've implemented each decoder -- ***I have only tested Auth0***.

I'll have to setup some accounts with Netlify and Firebase and Magic to test, but figured would see if this pattern is what you had in mind.

Can see that `AuthHeaders` no longer has the "ugly" switch.

```
export const decodeAuthToken = async ({
  type,
  event,
  context,
}: {
  type: SupportedAuthTypes
  event: APIGatewayProxyEvent
  context: LambdaContext
}): Promise<AuthToken> => {
  const authDecoder = createAuthDecoder(type)
  const decoded = await authDecoder.decodeToken({ type, event, context })

  return decoded
}
```
It creates a decoder, call its decodeToken and returns.

The decoder (auth0, netlify, firebase) does its own decoding or raises an exception.

### Issues

* Have only tested Auth0 at moment
* firebase/magicLink decoders have duplicate code (as do netlify and goTrue)
* not sure about the custom implementation
* there's no more "default" it totally unsupported auth type case
* I didn't set the decoder in create decoder (like client does). Wondering if my AuthType keys will all match.
* am new to TypeScript so fingers crossed

Reply:
Ok! I've left some feedback!